src/erasure-code/ErasureCode.cc

```diff
@@ -332,17 +332,40 @@ int ErasureCode::decode_concat(const map<int, bufferlist> &chunks,
 			       bufferlist *decoded)
 {
   set<int> want_to_read;
+  set<int> decode_chunks;
+  bool need_decode = false;
 
   for (unsigned int i = 0; i < get_data_chunk_count(); i++) {
     want_to_read.insert(chunk_index(i));
   }
+  if (chunks.size() < get_data_chunk_count()) {
+    // for partial_read
+    for(map<int, bufferlist>::const_iterator i = chunks.begin();
+       i != chunks.end();
+       ++i) {
+      if (want_to_read.find(i->first) == want_to_read.end()) {
+        need_decode = true;
+        break;
+      }
+      decode_chunks.insert(i->first);
+    }
+    if (!need_decode) {
+      want_to_read.swap(decode_chunks);
+    }
+  }
```

Comment:
Suggested:

  set<int> want_to_read;
  for (unsigned int i = 0; i < get_data_chunk_count(); i++) {
    want_to_read.insert(chunk_index(i));
  }

  bool need_decode = false;
  set<int> decode_chunks;
  if (chunks.size() < get_data_chunk_count()) {

    // for partial_read
    for (const auto& [thing_key, thing_bl] : chunks) {
      // note that we do use c++20
      if (want_to_read.contains(thing_key])) {
        decode_chunks.insert(i->first);
      } else {
        need_decode = true;
        break;
      }
    }

    if (!need_decode) {
      want_to_read.swap(decode_chunks);
    }

Reply:
Nice cleanup!